File: src/BrightLocal/Batch.php
<?php
namespace BrightLocal;

use BrightLocal\Exceptions\BatchAddJobException;
use BrightLocal\Exceptions\BatchCommitException;
use BrightLocal\Exceptions\BatchCreateException;
use BrightLocal\Exceptions\BatchDeleteException;

class Batch {

    private Api $api;
    private int $batchId;

    public function __construct(Api $api) {
        $this->api = $api;
    }

    public function setId(int $batchId) {
        $this->batchId = $batchId;
    }

    public function getId(): int {
        return $this->batchId;
    }

    /**
     * @throws BatchCreateException
     */
    public function create(bool $stopOnJobError = false, ?string $callbackUrl = null): Batch {
        $params = ['stop-on-job-error' => (int) $stopOnJobError];
        if (!empty($callbackUrl)) {
            $params['callback'] = $callbackUrl;
        }
        $response = $this->api->post('/v4/batch', $params);
        if (!$response->isSuccess() || (isset($response->getResult()['batch-id']) && !is_int($response->getResult()['batch-id']))) {
            throw new BatchCreateException(sprintf('An error occurred and we were unable to create the batch. Errors:%s', print_r($response->getResult()['errors'], true)));
        }
        $this->setId((int) $response->getResult()['batch-id']);
        return $this;
    }

    /**
     * @throws BatchCommitException
     */
    public function commit(): bool {
        $response = $this->api->put('/v4/batch', [
            'batch-id' => $this->batchId
        ]);
        if (!$response->isSuccess()) {
            throw new BatchCommitException(sprintf('An error occurred and we aren\'t able to commit the batch. Errors:%s', print_r($response->getResult()['errors'], true)));
        }
        return $response->isSuccess();
    }

    /**
     * @throws BatchAddJobException
     */
    public function addJob(string $resource, array $params = []): BatchResponse {
        $params['batch-id'] = $this->batchId;
        $response = $this->api->post($resource, $params);
        if (!$response->isSuccess()) {
            throw new BatchAddJobException(sprintf('An error occurred and we weren\'t able to add the job to the batch. Errors:%s', print_r($response->getResult()['errors'], true)));
        }
        return $response;
    }

    /**
     * @throws BatchDeleteException
     */
    public function delete(): bool {
        $response = $this->api->delete('/v4/batch', [
            'batch-id' => $this->batchId
        ]);
        if (!$response->isSuccess()) {
            throw new BatchDeleteException(sprintf('An error occurred and we weren\'t able to delete the batch. Errors:%s', print_r($response->getResult()['errors'], true)));
        }
        return $response->isSuccess();
    }

    public function getResults(): BatchResponse {
        return $this->api->get('/v4/batch', [
            'batch-id' => $this->batchId
        ]);
    }
}

File: src/BrightLocal/batch.py
from typing import Optional
from brightlocal_api import ApiResponse


class Batch:

    def __init__(self, api, batch_id: Optional[int] = None):
        self.api = api
        self.batch_id = batch_id

    def get_id(self) -> Optional[int]:
        return self.batch_id

    def commit(self) -> bool:
        response = self.api._make_request('/v4/batch', {'batch-id': self.batch_id}, 'PUT')
        if not response.isSuccess():
            raise Exception('An error occurred and we aren\'t able to commit the batch.')
        return True

    def add_job(self, resource: str, params: dict) -> ApiResponse:
        params['batch-id'] = self.batch_id
        response = self.api._make_request(resource, params)
        if not response.isSuccess():
            raise Exception('An error occurred and we weren\'t able to add the job to the batch.')
        return response

    def delete(self) -> bool:
        response = self.api._make_request('/v4/batch', {'batch-id': self.batch_id}, 'DELETE')
        if not response.isSuccess():
            raise Exception('An error occurred and we weren\'t able to delete the batch.')
        return True

    def stop(self) -> bool:
        response = self.api._make_request('/v4/batch/stop', {'batch-id': self.batch_id}, 'PUT')
        if not response.isSuccess():
            raise Exception('An error occurred and we weren\'t able to stop the batch.')
        return True

    def get_results(self) -> ApiResponse:
        response = self.api._make_request('/v4/batch', {'batch-id': self.batch_id})
        if not response.isSuccess():
            raise Exception('An error occurred and we weren\'t able to find the batch.')
        return response
